### PR TITLE
docs: Use Sphinx for generating API docs.

### PR DESCRIPTION
**Pull request by yaoyuannnn**
This enables automatic Python/C++ API documentation generation using Doxygen, Sphinx
and two Sphinx extensions, Breathe and Exhale. The automatic C++ docs generator
executes Doxygen to generate XML outputs, which will be parsed by Breathe/Exhale
into the Sphinx domain.

To build the htmls, go to `docs` and run `make html`. The htmls will be
generated in the `docs/build` folder.

### CODE REVIEW

File: docs/make.bat [hunk 1]
```diff
@@ -0,0 +1,35 @@
+@ECHO OFF
```
Comment (xyzsam) [inline docs/make.bat:1]: Why do we need a .bat file? I never intend to generate this on Windows :)

File: docs/Makefile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+SOURCEDIR     = source
+BUILDDIR      = build
+
+# Put it first so that "make" without argument is like "make help".
+help:
+	@$(SPHINXBUILD) -M help "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O)
+
+.PHONY: help Makefile
+
+# Catch-all target: route all unknown targets to Sphinx using the new
```
Comment (xyzsam) [inline docs/Makefile:17]: So what am I supposed to type to build?

File: docs/Makefile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+#
+
+# You can set these variables from the command line, and also
+# from the environment for the first two.
+SPHINXOPTS    ?=
+SPHINXBUILD   ?= sphinx-build
+SOURCEDIR     = source
+BUILDDIR      = build
+
+# Put it first so that "make" without argument is like "make help".
```
Comment (xyzsam) [inline docs/Makefile:11]: Why do we want this behavior?

File: docs/source/conf.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+autodoc_default_flags = ['members']
+autosummary_generate = True
+
+# Add any paths that contain templates here, relative to this directory.
+templates_path = ['_templates']
+
+# List of patterns, relative to source directory, that match files and
+# directories to ignore when looking for source files.
+# This pattern also affects html_static_path and html_extra_path.
+exclude_patterns = ['**test**', '**example**']
```
Comment (xyzsam) [inline docs/source/conf.py:98]: Should we clear this out?

**Comment by xyzsam** [comment]
Also, you've moved the location of the html output for Doxygen, which will break the auto-deployment setup. Can you update the path in config.yml so that this will continue to work until we can test and iron out the Sphinx docs auto-deployment?

File: docs/make.bat (hunk 1 above)
Comment (yaoyuannnn) [reply docs/make.bat:1]: This is added by sphinx-quickstart :) It generates a two Makefiles and a conf.py. Let me get rid of this.

File: docs/Makefile (hunk 2 above)
Comment (yaoyuannnn) [reply docs/Makefile:17]: `make html` will build everything and generate htmls.

File: docs/source/conf.py (hunk 4 above)
Comment (yaoyuannnn) [reply docs/source/conf.py:98]: Yes. Cleared this.

File: docs/Makefile (hunk 2 above)
Comment (xyzsam) [reply docs/Makefile:17]: Then why not make an `html` target instead? That's much more clear than a catch-all target, unless I'm missing something here.

File: docs/Makefile (hunk 3 above)
Comment (yaoyuannnn) [reply docs/Makefile:11]: I guess Sphinx-quickstart wants this. Let's keep this.

File: docs/Makefile (hunk 2 above)
Comment (yaoyuannnn) [reply docs/Makefile:17]: Sure. Added an `html` target.

**Comment by yaoyuannnn** [comment]
> Also, you've moved the location of the html output for Doxygen, which will break the auto-deployment setup. Can you update the path in config.yml so that this will continue to work until we can test and iron out the Sphinx docs auto-deployment?

Yes. Done.

**Comment by yaoyuannnn** [comment]
> This only changes the location where html is generated, but doesn't actually generate any Sphinx output yet right? I don't see any `make html` command yet.

Correct. This still generates the Doxygen htmls. I'll generate the Sphinx output int the next commit.

**Pull request merged**